Assets/MixedRealityToolkit/MSBuild/Scripts/CompilationPlatformInfo.cs

```diff
@@ -0,0 +1,176 @@
+
+namespace Microsoft.MixedReality.Toolkit.MSBuild
+{
+    /// <summary>
+    /// This contains parsed information using Unity configuration about a specific compilation platform.
+    /// </summary>
+    public class CompilationPlatformInfo
+    {
+        /// <summary>
+        /// Given a non-editor <see cref="AssemblyDefinitionPlatform"/> platform, creates an instances of CompilationPlatform fetching defines and references.
+        /// </summary>
+        /// <param name="platform">The platform to use for parsing.</param>
+        /// <returns>The <see cref="CompilationPlatformInfo"/> containing building infromation for the platform.</returns>
+        public static CompilationPlatformInfo GetCompilationPlatform(AssemblyDefinitionPlatform platform)
+        {
+            if (platform.BuildTarget == BuildTarget.NoTarget)
+            {
+                throw new ArgumentException(nameof(platform), "Provided an editor platform, use GetEditorPlatform.");
+            }
+
+            AssemblyBuilder builder = new AssemblyBuilder("dummy.dll", new string[] { @"Editor\dummy.cs" })
+            {
+                buildTarget = platform.BuildTarget,
+                buildTargetGroup = Utilities.GetBuildTargetGroup(platform.BuildTarget),
+                flags = AssemblyBuilderFlags.None
+            };
+
+            HashSet<string> platformCommonDefines = new HashSet<string>(builder.defaultDefines);
+            HashSet<string> playerDefines = new HashSet<string>(builder.defaultDefines);
+
+            HashSet<string> platformCommonReferences = new HashSet<string>(FilterOutProjectReferences(builder.defaultReferences));
+            HashSet<string> playerReferences = new HashSet<string>(FilterOutProjectReferences(builder.defaultReferences));
+
+            builder.flags = AssemblyBuilderFlags.EditorAssembly;
+            HashSet<string> inEditorDefines = new HashSet<string>(builder.defaultDefines);
+            platformCommonDefines.RemoveWhere(t => !inEditorDefines.Contains(t));
+
+            HashSet<string> inEditorReferences = new HashSet<string>(FilterOutProjectReferences(builder.defaultReferences));
+            platformCommonReferences.RemoveWhere(t => !inEditorReferences.Contains(t));
+
+            // Remove the common
+            playerDefines.RemoveWhere(t => platformCommonDefines.Contains(t));
+            inEditorDefines.RemoveWhere(t => platformCommonDefines.Contains(t));
+
+            playerReferences.RemoveWhere(t => platformCommonReferences.Contains(t));
+            inEditorReferences.RemoveWhere(t => platformCommonReferences.Contains(t));
+
+            // Get specialized
+            playerDefines.RemoveWhere(t => platformCommonDefines.Contains(t));
+            inEditorDefines.RemoveWhere(t => platformCommonDefines.Contains(t));
+
+            playerReferences.RemoveWhere(t => platformCommonReferences.Contains(t));
+            inEditorReferences.RemoveWhere(t => platformCommonReferences.Contains(t));
+
```

Comment:
Nit: For future cases, consider using things like Except, Intersect, Union and the corresponding HashSet operating functions:

[https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1.exceptwith?view=netframework-4.8](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1.exceptwith?view=netframework-4.8)

If only because it can be at times more readable to read the fundamental set operation words (i.e. this thing  = A intersect B) instead of going A but remove everything from one and then do the same to the other and then use the stuff that remains.